odl/space/fspace.py

```diff
@@ -1141,26 +931,334 @@ def __init__(self, fspace, fcall):
         fspace : `FunctionSpace`
             Set of functions this element lives in.
         fcall : callable
-            The actual instruction for out-of-place evaluation.
-            It must return an `FunctionSet.range` element or a
-            ``numpy.ndarray`` of such (vectorized call).
+            Object used to evaluate the function. Must support
+            vectorization and accept a sequence of
+            coordinate arrays ``x[0], ..., x[d]`` in sparse or dense
+            form, and return (or write to the ``out`` array) an
+            array of appropriate shape.
+        """
+        super().__init__(fspace)
+        self._call_has_out, self._call_out_optional = _fcall_out_type(fcall)
+
+        if not self._call_has_out:
+            # Out-of-place-only
+            decorator = preload_first_arg(self, 'in-place')
+            self._call_in_place = decorator(_default_in_place)
+            self._call_out_of_place = fcall
+        elif self._call_out_optional:
+            # Dual-use
+            self._call_in_place = self._call_out_of_place = fcall
+        else:
+            # In-place-only
+            self._call_in_place = fcall
+            decorator = preload_first_arg(self, 'out-of-place')
+            self._call_out_of_place = decorator(_default_out_of_place)
+
+    @property
+    def domain(self):
+        """Set of objects on which this function can be evaluated."""
+        return self.space.domain
+
+    @property
+    def out_dtype(self):
+        """Output data type of this function.
+
+        If ``None``, the output data type is not uniquely pre-defined.
+        """
+        return self.space.out_dtype
+
+    @property
+    def scalar_out_dtype(self):
+        """Scalar variant of ``out_dtype`` in case it has a shape."""
+        return self.space.scalar_out_dtype
+
+    @property
+    def out_shape(self):
+        """Shape of function values, ``()`` for scalar output."""
+        return self.space.out_shape
+
+    @property
+    def tensor_valued(self):
+        """``True`` if the output is multi-dim. output, else ``False``."""
+        return self.space.tensor_valued
+
+    def _call(self, x, out=None, **kwargs):
+        """Raw evaluation method."""
+        if out is None:
+            return self._call_out_of_place(x, **kwargs)
+        else:
+            self._call_in_place(x, out=out, **kwargs)
+
+    def __call__(self, x, out=None, **kwargs):
```

Comment:
Agreed. The writing of this  was also mostly incremental, fixing one failing test at a time.